pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+// for maintaining a list of service endpoints used by this Pilot.
+//
+// Controllers under https://github.com/istio/istio/tree/master/pilot/platform
+// update the list of service endpoints in this Mesh with those
+// available in a platform specific service registry.
+//
+// Under the hoods, the Mesh implements the necessary logic required for endpoint
+// discovery based on routing rules and endpoint subsets.
+// See https://github.com/istio/api/search?q=in%3Afile+"message+Subset"+language%3Aproto
+// This logic includes comparing the updated list provided by Controllers with
+// what this view holds and accordingly updating internal structures used for
+// endpoint discovery.
+type Mesh struct {
+
+	// allEndpoints is a map of the UID to the server instance. The UID is expected
+	// to be unique across the mesh. Mesh uses this map during updates from
+	// Controllers to locate the previous representation  of the endpoint and check
+	// for changes to its metadata that could result in changes to the subsetEndpoints.
+	allEndpoints map[string]*Endpoint
+
+	// subsetDefinitions holds the metadata, basically the set of lable name values
+	// that ought to match corresponding attributes of the endpoint.
+	// The actual subset of allEndpoints that match these subset definitions is
+	// stored in subsetEndpoints. Mesh uses this map for identifying config changes
+	// to the subset's metadata. The key of this map is the Subset name.
+	subsetDefinitions map[string]*route.Subset
+
+	// subsetEndpoints allows for look-ups from the Subset name to a set of Endpoint UIDs.
+	// The UIDs in the set are guaranteed to exist in allEndpoints and the
+	// corresponding LbEndpointis guaranteed to satisfy the labels in the corresponding
+	// Subset.
+	subsetEndpoints subsetEndpoints
+
+	// reverseAttrMap provides reverse lookup from label name > label value >
+	// endpointUIDSet. All Endpoints in the set are guaranteed to match the
+	// attribute with the label name and value. Mesh uses the map for quickly
+	// updating subsetEndpoints in response to subset label changes.
+	reverseAttrMap attributeValues
+
+	// reverseEpSubsets provides a reverse lookup from endpoint key to the set of
+	// subsets this endpoint matched to. All subset keys associated with the endpoint
+	// are guaranteed to exist in subsetDefinitions. Mesh uses this to reverse
+	// lookup all subsets impacted by changes to Endpoint metadata.
+	reverseEpSubsets endpointSubsets
+
+	// Mutex guards guarantees consistency of updates to members shared across
+	// threads.
+	mu sync.RWMutex
+}
+
+// Endpoint is a platform independent representation of a Mesh Endpoints that
+// uses Envoy v2 API's LbEndpoint as its internal implementation. It also
+// provides utility methods intended for platform specific Controllers.
+type Endpoint xdsapi.LbEndpoint
+
+// endpointUIDSet is a unique set of Endpoint UIDs.
+type endpointUIDSet map[string]bool
```

Comment:
Key ? How is it used ?

Reply:
I was hoping https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html would explain UID and this is a set ... apologies for my golang ignorance .. but aren't sets supposed to be map[whatever]bool?